### ISSUE TITLE
ptop crash

### ISSUE THREAD

**Issue by walles** [walles/px#67]
On mac, inside `ptop`:

```
  File “.bootstrap/pex/pex.py”, line 352, in execute
 File “.bootstrap/pex/pex.py”, line 284, in _wrap_coverage
 File “.bootstrap/pex/pex.py”, line 315, in _wrap_profiling
 File “.bootstrap/pex/pex.py”, line 397, in _execute
 File “.bootstrap/pex/pex.py”, line 495, in execute_entry
 File “.bootstrap/pex/pex.py”, line 502, in execute_module
 File “/usr/local/Cellar/python@2/2.7.15/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py”, line 192, in run_module
   fname, loader, pkg_name)
 File “/usr/local/Cellar/python@2/2.7.15/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py”, line 72, in _run_code
   exec code in run_globals
 File “/Users/ssa/.pex/install/pxpx-1.0.27-py2.py3-none-any.whl.6a9a8fcab886cda24230300318bd0a9ba654ae93/pxpx-1.0.27-py2.py3-none-any.whl/px/px.py”, line 127, in <module>
   main()
 File “/Users/ssa/.pex/install/pxpx-1.0.27-py2.py3-none-any.whl.6a9a8fcab886cda24230300318bd0a9ba654ae93/pxpx-1.0.27-py2.py3-none-any.whl/px/px.py”, line 63, in main
   _main(sys.argv)
 File “/Users/ssa/.pex/install/pxpx-1.0.27-py2.py3-none-any.whl.6a9a8fcab886cda24230300318bd0a9ba654ae93/pxpx-1.0.27-py2.py3-none-any.whl/px/px.py”, line 88, in _main
   px_top.top()
 File “/Users/ssa/.pex/install/pxpx-1.0.27-py2.py3-none-any.whl.6a9a8fcab886cda24230300318bd0a9ba654ae93/pxpx-1.0.27-py2.py3-none-any.whl/px/px_top.py”, line 559, in top
   _top()
 File “/Users/ssa/.pex/install/pxpx-1.0.27-py2.py3-none-any.whl.6a9a8fcab886cda24230300318bd0a9ba654ae93/pxpx-1.0.27-py2.py3-none-any.whl/px/px_top.py”, line 540, in _top
   current = px_process.get_all()
 File “/Users/ssa/.pex/install/pxpx-1.0.27-py2.py3-none-any.whl.6a9a8fcab886cda24230300318bd0a9ba654ae93/pxpx-1.0.27-py2.py3-none-any.whl/px/px_process.py”, line 301, in get_all
   ps_lines = call_ps()
 File “/Users/ssa/.pex/install/pxpx-1.0.27-py2.py3-none-any.whl.6a9a8fcab886cda24230300318bd0a9ba654ae93/pxpx-1.0.27-py2.py3-none-any.whl/px/px_process.py”, line 182, in call_ps
   env=env)
 File “/usr/local/Cellar/python@2/2.7.15/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py”, line 394, in __init__
   errread, errwrite)
 File “/usr/local/Cellar/python@2/2.7.15/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py”, line 1033, in _execute_child
   _close_in_parent(c2pwrite)
 File “/usr/local/Cellar/python@2/2.7.15/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py”, line 922, in _close_in_parent
   os.close(fd)
OSError: [Errno 9] Bad file descriptor
```

**Comment by walles** [comment]
This was while pressing `ENTER` (to show process info) and `CTRL-C` (to interrupt `lsof`).

Might be reproducable.

**Comment by walles** [comment]
Maybe related to this?
https://stackoverflow.com/questions/7686275/what-can-lead-to-ioerror-errno-9-bad-file-descriptor-during-os-system

**Comment by walles** [comment]
Or to this? Same thing reported in two different places:
* https://stackoverflow.com/a/18454553/473672
* https://bugs.python.org/issue18851

Need to know the Python version used to know if this was the issue.

**Comment by walles** [comment]
From the stack trace, it was 2.7.15.

**Comment by walles** [comment]
This has happened to others: https://www.google.com/search?q="_close_in_parent"+"bad+file+descriptor".

Need to look in the Python sources for this; it might be some race.

**Comment by walles** [comment]
Lots of LOC: https://github.com/python/cpython/blob/ca079a3ea30098aff3197c559a0e32d42dda6d84/Lib/subprocess.py#L1033

**Comment by walles** [comment]
Let's try a timeline:
* User presses Enter to start paging process info for some process.
* A pager is started by `page_process_info()` using `subprocess.Popen()`
* Its `stdin` `fileno` is passed to a background thread for pumping process info into.
  * When done, the pumping thread will `os.close()` that `fileno`.
  * NOTE: Could this `fileno` go out of scope and be garbage collected / re-used before `page_process_info()` returns? Leaving the thread pumping stuff into an invalid `fileno`?
    * Maybe wait for the background thread to finish before returning from `page_process_info()`?
* `page_process_info()` finishes and ordinary `ptop` operations resume.
* We get the exception reported here.

**Comment by walles** [comment]
I can repro!

* set pager to `less`
* run `ptop`
* press `enter` on some process
* watch `less` appear
* press `q`
* wait for `less` to react
* when you're back in `ptop`: Boom!

Unfortunately this repro is pretty sucky, I tried it about 20 times and it crashed only once. But still...
